libraries/chain/include/eosio/chain/wasm_interface_private.hpp

```diff
@@ -84,48 +124,90 @@ namespace eosio { namespace chain {
+         return (eosvmoc || wasm_runtime_time == wasm_interface::vm_type::eos_vm_oc);
+      }
+#endif
+
+      const std::unique_ptr<wasm_instantiated_module_interface>& get_instantiated_module(
+         const digest_type&   code_hash,
+         const uint8_t&       vm_type,
+         const uint8_t&       vm_version,
+         transaction_context& trx_context)
       {
-         wasm_cache_index::iterator it = wasm_instantiation_cache.find(
-                                             boost::make_tuple(code_hash, vm_type, vm_version) );
-         const code_object* codeobject = nullptr;
-         if(it == wasm_instantiation_cache.end()) {
-            codeobject = &db.get<code_object,by_code_hash>(boost::make_tuple(code_hash, vm_type, vm_version));
-
-            it = wasm_instantiation_cache.emplace( wasm_interface_impl::wasm_cache_entry{
-                                                      .code_hash = code_hash,
-                                                      .last_block_num_used = UINT32_MAX,
-                                                      .module = nullptr,
-                                                      .vm_type = vm_type,
-                                                      .vm_version = vm_version
-                                                   } ).first;
+         wasm_cache_index::iterator it;
+         if (trx_context.control.is_write_window()) {
+            // When in write window (either read only threads are not enabled or
+            // they are not schedued to run), only main thread is processing
+            // transactions. No need to lock.
+            it = wasm_instantiation_cache.find( boost::make_tuple(code_hash, vm_type, vm_version) );
+         } else {
+            std::shared_lock g(instantiation_cache_mutex);
+            it = wasm_instantiation_cache.find( boost::make_tuple(code_hash, vm_type, vm_version) );
          }
 
-         if(!it->module) {
-            if(!codeobject)
-               codeobject = &db.get<code_object,by_code_hash>(boost::make_tuple(code_hash, vm_type, vm_version));
+         if (it != wasm_instantiation_cache.end()) {
+            // An instantiated module's module should never be null.
+            assert(it->module);
+            return it->module;
+         }
 
-            auto timer_pause = fc::make_scoped_exit([&](){
-               trx_context.resume_billing_timer();
-            });
-            trx_context.pause_billing_timer();
-            wasm_instantiation_cache.modify(it, [&](auto& c) {
-               c.module = runtime_interface->instantiate_module(codeobject->code.data(), codeobject->code.size(), code_hash, vm_type, vm_version);
-            });
+         if (trx_context.control.is_write_window()) {
+            return build_and_get_instantiated_module(code_hash, vm_type, vm_version, trx_context);
+         } else {
+            // Use unique lock as we need to modify the cache
+            std::unique_lock g(instantiation_cache_mutex);
```

Comment:
You can use a `std::scope_lock` or `std::lock_guard` here instead. You only need a `unique_lock` if you are going to manually `unlock` it.